### PR TITLE
maxSliceColumns in fromRValues

### PR DESCRIPTION
**Pull request by rintcius**
First part of #qz-3823

### CODE REVIEW

File: yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala [hunk 1]
```diff
@@ -1787,92 +1787,94 @@ object Slice {
...
         val ref = ColumnRef(cpath, (cvalue.cType))
 
         val updatedColumn: ArrayColumn[_] = cvalue match {
           case CBoolean(b) =>
-            acc.getOrElse(ref, ArrayBoolColumn.empty()).asInstanceOf[ArrayBoolColumn].unsafeTap { c =>
-              c.update(sliceIndex, b)
-            }
+            val c = acc.getOrElse(ref, ArrayBoolColumn.empty()).asInstanceOf[ArrayBoolColumn]
+            c.update(sliceIndex, b)
+            c
```
Comment (rintcius) [inline yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala]: I removed `unsafeTap` here (see https://github.com/scalaz/scalaz/issues/1121), but I am tempted to get rid of `update` returning Unit. Any reason we shouldn't get rid of this?

**Comment by djspiewak** [comment]
It's worth noting that, because `maxSliceBytes` is only counting *defined* values at present, it's actually not going to catch sparse slices at all! So this doesn't really solve the problem.

File: yggdrasil/src/main/scala/quasar/yggdrasil/table/ArrayColumn.scala [hunk 2]
```diff
@@ -65,6 +68,8 @@ class ArrayBoolColumn(val defined: BitSet, val values: BitSet) extends ArrayColu
     defined.set(row)
     if (value) values.set(row) else values.clear(row)
   }
+
+  def byteSize: Long = defined.cardinality * ByteSize.boolean
```
[the remark this answers is not in the document]
Comment (rintcius) [reply yggdrasil/src/main/scala/quasar/yggdrasil/table/ArrayColumn.scala]: It's certainly very low, it's 0 for `boolean` (because I took `ByteSize.boolean` to be 0 for now) :) Accuracy is not the goal in this first pass, see https://github.com/quasar-analytics/quasar/pull/3602/files#diff-ed9d0666d3f82bb8a4d79fd56d9827e9R24 
I was thinking we could make it more accurate after we are convinced that we should take this path _and_ we have more clarity at whether more accuracy will gain us enough and to what extent it's worth the extra CPU time, i.e. where should we live on the accuracy vs CPU time spectrum. I picked "minimize CPU time" on this spectrum for now and ignored all the (fairly) fixed costs like `definedness` and array initialization.

Personally I'd like to make these calculations much more accurate, but I haven't built enough intuition to be sure it's worth it.

**Comment by rintcius** [comment]
> It's worth noting that, because maxSliceBytes is only counting defined values at present, it's actually not going to catch sparse slices at all! So this doesn't really solve the problem.

@djspiewak yeah, that's unfortunate. Until @alissapajer pinged me yesterday mentioning qz-3748 , I wasn't aware that slice sparsity was the problem to be solved. I thought the problem to be solved was improve the counting rows strategy as a way to determine whether slices fit in memory.

File: mimir/src/main/scala/quasar/mimir/ReductionLib.scala [hunk 3]
```diff
@@ -676,7 +676,7 @@ trait ReductionLibModule[M[+ _]] extends ColumnarTableLibModule[M] {
       }
 
       def extract(res: Result): Table =
-        Table.fromRValues(res.toStream, None)
+        Table.fromRValues0(res.toVector)
```
[the remark this answers is not in the document]
Comment (rintcius) [reply mimir/src/main/scala/quasar/mimir/ReductionLib.scala]: > I don't see any reason to use scala.Stream

not sure what you mean here, I changed `values` from `Stream` to `Vector` like this. So, I guess I agree? 

```
    def fromRValues(values: Vector[RValue], maxSliceBytes: Long): Table
    def fromRValues0(values: Vector[RValue]): Table = fromRValues(values, Config.maxSliceBytes)
```

> Why the 0? Why shouldn't this just be the default for fromRValues? 

I created 2 methods because then `TableModule` is the only one that needs to know about the default for `maxSliceBytes` being `Config.maxSliceBytes`, so `ColumnarTableModule` doesn't need to duplicate this.
And I want to have `maxSliceBytes: Long` available as an argument, for more flexibility (eg calling directly from tests without having to fiddle with the Config). Tying in the `Config` is a separate concern.

**Comment by alissapajer** [comment]
@rintcius yeah, so the problem is still to improve counting rows strategy to determine if slices fit into memory, we just also need to take into account maximally sparse data as the worst case scenario. And it's only a worst case scenario because, in the column implementation, we allocate an `Array[A]` for each column, so even if only one value in that array is defined, we've allocated a potentially large array.

I was excited that we can do better than just setting a `maxNumOfColumns` by calculating the size of these allocated arrays, and then having slices that are close to the "perfect" size each time. But since that is a more complicated calculation, as a first pass we could just add a `maxNumOfColumns`. In that case, if we're dealing with relatively dense data, we'll hit the `maxNumOfRows` bound first and cut a new slice then. But if we're dealing with very sparse data, we'll hit the `maxNumOfColumns` bound first and cut then.

I do think that a calculation-based approach is correct here no matter what bounds we choose, meaning that we want to set bounds according to a target slice size. And then we can document this target as well.

Unless we inspect variable-sized values (like strings) as we construct slices, it will still be possible to crash slamengine if a dataset has huge values. Crashing the application is a big deal if it's running as a server with many users. But there is also an overhead to calculating the size of each value before putting it into a slice.

@djspiewak from the business perspective, we should clarify what problems we need to solve now versus what problems we need to solve later. I see two problems, both of which create slices that are larger than can fit into memory, crashing slamengine. @rintcius do you see any other problems?

1) Sparse data can create large value arrays because we allocate an array with size equal to the height of the column.
2) Variable-sized data can create large value arrays because each individual value is large.

Even with fixed-sized data, some data (`Boolean`) are smaller than other data (`Long`) so we can optimize slice size using that information. But we could consider that more of an optimization, or solve it in the same pass as when we fix # 1.

File: mimir/src/main/scala/quasar/mimir/ReductionLib.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (djspiewak) [reply mimir/src/main/scala/quasar/mimir/ReductionLib.scala]: Why not use default parameters? Like `maxSliceBytes: Option[Long] = None`

**Comment by djspiewak** [comment]
@alissapajer I'm not sure I understand your point 2. What do you mean by "variable-sized data"?

I see really two problems in play here that we need to solve: 

- Columns are very large contiguous chunks of memory and we can't have too many of them in memory at once
- We pervasively make assumptions about how the `columns` map is relatively small and amenable to inefficient O(n) (and worse!) traversals

Sparse data in general is very inefficient and can push on the first problem, but it's really only when that sparse data has a large set of row-specific fields (so, it's very sparse *and* it's very very wide) that we start to see both problems together. So basically, document-oriented data.

The interesting thing about the approach being experimented with in this PR is the fact that it only addresses the first problem at a conceptual level, even assuming it were implemented to correctly count the space impact of sparse columns.  We could theoretically build a slice which is well within the size bound but still has far too many columns to be practically manipulated by some of our functions (e.g. look at the implementation of `OuterObjectConcat`).

I guess there's probably a more general cost model here that we are feeling our way towards. Memory footprint is certainly part of it, as is column size, as is row size, as is quantization to power-of-2 sizes on both (especially rows), etc.

**Comment by alissapajer** [comment]
Agreed that we're feeling our way toward a more generalized cost model. For the first step here we'll have to figure out how far exactly to generalize.

@djspiewak point 2. is the situation where someone has 100MB string values in their data.

File: mimir/src/main/scala/quasar/mimir/ReductionLib.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (rintcius) [reply mimir/src/main/scala/quasar/mimir/ReductionLib.scala]: > Why not use default parameters? Like maxSliceBytes: Option[Long] = None

Because that would shift the knowledge about `Config` from the base trait (`TableModule`) to its children (eg `ColumnarTableModule`)

**Comment by djspiewak** [comment]
@alissapajer Yeah point 2 is insanely tricky. There are some JVM internals we can use to get approximates information, but it's a pain and it can be slow if not done carefully. Remember this is all happening at the row level; it *needs* to be very very fast.

File: mimir/src/main/scala/quasar/mimir/ReductionLib.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (djspiewak) [reply mimir/src/main/scala/quasar/mimir/ReductionLib.scala]: Isn't the `Config` already known to the children?

[the remark this answers is not in the document]
Comment (rintcius) [reply mimir/src/main/scala/quasar/mimir/ReductionLib.scala]: Maybe, I just tend to move these things outside as far as possible. 

I just changed it (improved it from my perspective) because I needed to touch the methods anyway.

[the remark this answers is not in the document]
Comment (djspiewak) [reply mimir/src/main/scala/quasar/mimir/ReductionLib.scala]: I do like driving things to the outside, I just don't like the `0` variant. Not a big deal, just something that maybe we can improve later as we get further away from the cake.

**Comment by alissapajer** [comment]
Regarding point 1: 
I agree that we'll want these `Config` parameters as variables that we can pass in. So as not to conflate concerns, we can do that in a separate pass (in a separate PR). Also, I think that we can hold off on doing that until we are at the development stage of implementing a feature that will account for various characteristics of the table; that way we won't be over-engineering and we'll know more exactly at which stage we'll be acquiring the table characteristics and so will have a better sense of how to pass them around and pass them in.

Regarding point 2:
If a connector already knows the length of strings it's storing, then we could definitely take advantage of that to better create our slices. And so if we exposed `char` and `varchar` we could then hang on to the string length information and pass it along to mimir.

I'm not sure if we know fully yet what properties our customer's data will have initially (or really, which properties we want to support versus which we don't). I assume (@djspiewak ? @beckyconning ?) that we need to support document-oriented data (which results in sparse datasets with lots of columns), which is why we've prioritized fixing that case now.

**Comment by alissapajer** [comment]
Also, this is a general theme that we should keep in mind: if there is some information that would make slamengine more performant, and it's expensive for us to calculate that, then we should consider optionally asking the connector for that information. The idea being, if that have that information, we use it, if they don't, we do the best we can.

This can be applied to things like table size, indexing, smart sampling, etc.

**Comment by rintcius** [comment]
@alissapajer @djspiewak 
I'm busy adding `maxSliceColumns`. I'm still inclined to leave the `maxSliceBytes` code in. 

Even with the current impl of this, I'd think that this would be a big improvement over just calculating the number of rows (although the name `maxSliceBytes` is misleading right now, since the real amount of memory in use would be higher than indicated here). 

WDYT? Do you think there's too big a chance `maxSliceBytes` code won't be useful in the end? Other reasons to remove it for now?

**Comment by djspiewak** [comment]
> Even with the current impl of this, I'd think that this would be a big improvement over just calculating the number of rows (although the name maxSliceBytes is misleading right now, since the real amount of memory in use would be higher than indicated here).

Except calculating the number of rows is *also* important. `maxSliceBytes` doesn't capture things like heap fragmentation effects, which become quite significant when arrays become this large. So this means that things like quantizing to powers of 2 and having hard caps on row count *even* if well below byte size is always going to be something that we need to do.

Additionally, `maxSliceBytes` as implemented ignores undefined values, which is inaccurate to the point of not actually being meaningful. Arrays take up space regardless of whether or not they have a non-default value, especially arrays of primitives. `maxSliceBytes` *needs* to take that into account to be in any way meaningful.

I'm not saying that byte-counting is necessarily a bad idea – in fact, I think it's quite a good one – but it needs to be done differently in order for it to be useful. Row and column counting will also *always* have a place in the limitation mechanism, and for now, row and column counting are probably an easier and more direct way of getting at the limitation we need.

Remember that sparse document-oriented data is just always going to be terrible in mimir. Our goal is to make it *not crash*. We don't care so much about being super-optimal beyond that. Other slice limits (such as slices which are dense, at the row limit, and overflowing memory) are definitely a concern, but I'm not as concerned about them *right now*. I just want us to not crash in this one case.

**Comment by rintcius** [comment]
> Except calculating the number of rows is also important. maxSliceBytes doesn't capture things like heap fragmentation effects, which become quite significant when arrays become this large. 

Row counting is still there though.

But yeah, I understand if it's not a priority right now. It would require more work to get this in properly. I'll remove `maxSliceBytes` from this PR.

File: yggdrasil/src/main/scala/quasar/yggdrasil/Config.scala [hunk 4]
```diff
@@ -19,8 +19,9 @@ package quasar.yggdrasil
 object Config {
   val idSource = new FreshAtomicIdSource
 
-  def hashJoins         = true
-  def maxSliceSize: Int = 20000
+  def hashJoins = true
+  def maxSliceSize: Int = 16384 // should be power of 2
```
[the remark this answers is not in the document]
Comment (alissapajer) [reply yggdrasil/src/main/scala/quasar/yggdrasil/Config.scala]: :+1: to the new name

File: yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala [hunk 5]
```diff
@@ -1883,25 +1889,62 @@ object Slice {
...
+        case jv +: xs =>
+          val columnRefValues: Vector[(ColumnRef, CValue)] =
+            jv.flattenWithPath.map { case (p, v) => (ColumnRef(p, v.cType), v) }
+          val nextNrColumns = (columnRefValues.map(_._1) ++ into.keys).toSet.size
+
+          // NB: don't allow exceeding maxColumns
+          // EXCEPT if all these columns come from a single RValue
+          // Later we can hopefully get rid of this exception
+          // (e.g. once we have table specific configs)
+          if ((nextNrColumns <= maxColumns) || (into.isEmpty)) {
```
[the remark this answers is not in the document]
Comment (djspiewak) [reply yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala]: @rintcius I agree that we never want to separate single `RValue`s across slices. I'm not sure how that would even result in a usable encoding of the data, so it seems like a good restriction. :-) If a *single* value just has a zillion columns, there's honestly not much we can do.

File: yggdrasil/src/main/scala/quasar/yggdrasil/Config.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (rintcius) [reply yggdrasil/src/main/scala/quasar/yggdrasil/Config.scala]: ok! I'll do that in the next PR (implementing `maxSliceColumns` for canonicalize)

**Comment by alissapajer** [comment]
@rintcius do you have any further work planned for this PR? Doing the `canonicalize` work in a second PR is fine, just want to make sure I review things when they're ready :smile:

**Comment by rintcius** [comment]
@alissapajer it's ready for review :)

**Comment by alissapajer** [comment]
Looks good, nice work! Didn't quite get to finish reviewing it all tonight, so I'm going to finish that tomorrow and then we can land this!

File: yggdrasil/src/test/scala/quasar/yggdrasil/table/SliceSpec.scala [hunk 6]
```diff
@@ -71,6 +71,94 @@ class SliceSpec extends Specification with ScalaCheck {
...
+
+    val slices = Slice.fromRValues(values, maxRows = Math.max(totalRows, 1), maxColumns = nrColumnsBiggestValue).toVector
+    assertSlices(values, slices, be_>(0))
+  }
+
+  def testFromRValuesMaxSliceColumnsLowerThanBiggestValue(values: Vector[CValue]) = {
+    val (totalRows, nrColumnsBiggestValue, totalColumns) = valueCalcs(values)
+
+    val slices = Slice.fromRValues(values, maxRows = Math.max(totalRows, 1), maxColumns = nrColumnsBiggestValue - 1).toVector
+    assertSlices(values, slices, be_>(0))
```
Comment (alissapajer) [inline yggdrasil/src/test/scala/quasar/yggdrasil/table/SliceSpec.scala]: Could we add another test that manually checks that when we have too many columns, a new slice gets cut at precisely that point (modulo the `RValue` boundary)?

For example, we input three `RValue`s, the first two of which, combined, have more columns than `maxColumns`. Then we could verify that the result has two slices. Or something along those lines.

This test checks that we can handle an `RValue` with more columns than `maxColumns`. It would also be good to verify that we're actually splitting slices at the correct boundaries.

File: yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala [hunk 7]
```diff
@@ -1883,25 +1889,62 @@ object Slice {
...
-      val (columns, size) = buildColArrays(values, Map.empty[ColumnRef, ArrayColumn[_]], 0)
+    def buildSlice(values: Vector[RValue]): (Slice, Vector[RValue]) = {
+      val (values0, values1) = values.splitAt(maxRows)
+      val (cs, sz, restValues) =
+        buildColArrays(values0, Map.empty[ColumnRef, ArrayColumn[_]], 0, values0.size)
+      val slice = new Slice {
+        val columns = cs
+        val size = sz
+      }
+      (slice, values1 ++ restValues)
```
Comment (alissapajer) [inline yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala]: `values1 ++ restValues` is changing the order of the `RValue`s. `values1` is the tail of `values` after the call `values.splitAt(maxRows)`. `restValues` were a part of `values0` and so they should be prepended to `values1`.

We should be able to verify this with a test. The unit tests you added do verify that the resulting slices have the elements in the correct order, so I think this case isn't being hit.

File: yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala [hunk 8]
```diff
@@ -1883,25 +1889,62 @@ object Slice {
     }
   }
 
-  def fromJValues(values: Stream[JValue]): Slice =
+  def fromJValues(values: Vector[JValue]): Stream[Slice] =
     fromRValues(values.flatMap(RValue.fromJValue))
 
-  def fromRValues(values: Stream[RValue]): Slice = {
-    val sliceSize = values.size
+  def fromRValues(
```
Comment (alissapajer) [inline yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala]: We'll need to improve the performance of this algorithm at some point. Currently, we take the max number of rows that we're allowed, then walk over them creating the column arrays until we have too many columns, or reach the end. If we have too many columns, then we stick the remaining rows back with the rest of the rows, and start over.

Ultimately we'll need a performant way to build slices from an `fs2.Stream` (since that's what lwcs provide). From an `fs2.Stream`, we can take as many rows/columns as we can, and then cut a slice when we hit the limit (whether it be from rows or columns). This current approach requires paging the maximum number of rows into memory before deciding if we have too many columns.

File: yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala (hunk 7 above)
Comment (rintcius) [reply yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala]: good catch, thanks!

File: yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala (hunk 8 above)
Comment (djspiewak) [reply yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala]: Unfortunately, the performance of this function as written is actually worse than it looks: all of the columns are allocated at the maximum row size even if the slice is cut off sooner! In other words, we're still eating a very similar memory footprint with very wide data, even with these changes.

The slice row bound estimation will also become an even more pressing issue as we move into direct ingestion of unparsed formats coming up from LWCs, since we won't have any way of knowing precisely how many rows we're looking at until we actually build the slices.

The solution here is to repeatedly re-estimate slice size and copy over data. Unless we *know* for an absolute fact that we're going to ingest up to the max row bound, it will be more efficient to follow a scheme like this:

- Allocate a default size column (suggestion: 32 rows is probably a good starting point, but we should eventually profile and figure it out) for each column that is necessary
- When a column is filled, double its size and reallocate, using `arraycopy` to move the data over
- Repeat until max row or column bound

Reminder: max row bound should be a power of two (probably 2^14 = 16384), *partially* for reasons of the above.

tldr: As written, this function is a slight improvement over what we're doing right now in that we don't allocate a zillion columns, but we're still using a *ton* more memory than we should be when we do hit the column bound. Exponential resize estimation as described above resolves this problem and opens the door to efficient boxless ingest from LWCs down the road.

Comment (rintcius) [reply yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala]: > Unfortunately, the performance of this function as written is actually worse than it looks: all of the columns are allocated at the maximum row size even if the slice is cut off sooner! 

I see the memory problem, but in what way is this a performance problem? Or do you just mean performance problems as a direct consequence of extra memory footprint? 

Would it help if we do the row reallocation _after_ slice creation? I.e. we would still initialize with `maxSliceRows` but when there are less actual rows we can scale it down by doing a single `arraycopy`. So only a temp. increase in memory footprint.

Or maybe a mix of these 2 approaches, if we need to decrease that amount of temp. memory footprint (eg start with an initial size of `maxSliceRows / 4`). Doing a final row reallocation after slice creation also could help in case number of rows is less than that default size (i.e. the 32 rows that was mentioned)

**Comment by rintcius** [comment]
Closing, in favor of #3644
